### ISSUE TITLE
Problem with big Zip64 files

### ISSUE THREAD

**Issue by MalteHillmann** [danifus/pyzipper#9]
Hello,

pyzipper has a problem with big Zip64 files. It always crashes with:
```
Traceback (most recent call last):
  File "[REDACTED].py", line 189, in <module>
    [REDACTED]
  File "[REDACTED]r.py", line 55, in [REDACTED]
    zipfile = pyzipper.AESZipFile([REDACTED])
  File "[REDACTED]\Python38-32\lib\site-packages\pyzipper\zipfile_aes.py", line 319, in __init__
    super().__init__(*args, **kwargs)
  File "[REDACTED]\Python38-32\lib\site-packages\pyzipper\zipfile.py", line 1775, in __init__
    raise e
  File "[REDACTED]\Python38-32\lib\site-packages\pyzipper\zipfile.py", line 1738, in __init__
    self._RealGetContents()
  File "[REDACTED]\Python38-32\lib\site-packages\pyzipper\zipfile.py", line 1861, in _RealGetContents
    x._decodeExtra()
  File "[REDACTED]\Python38-32\lib\site-packages\pyzipper\zipfile.py", line 785, in _decodeExtra
    extra_decoders[tp](ln, extra)
  File "[REDACTED]\Python38-32\lib\site-packages\pyzipper\zipfile.py", line 741, in decode_extra_zip64
    counts = struct.unpack(struct_str, extra[offset:offset+ln])
struct.error: unpack requires a buffer of 10 bytes
```

The zip file has a size of ~9,3 GiB. 
I found a possible fix by copying some of the lines of the current python3 [zipfile.py](https://github.com/python/cpython/blob/master/Lib/zipfile.py) to pyzippers [zipfile.py](https://github.com/danifus/pyzipper/blob/master/pyzipper/zipfile.py):

In pyzippers [zipfile.py](https://github.com/danifus/pyzipper/blob/master/pyzipper/zipfile.py) replace the lines 736 to 741:
```
            q_field_cnt = (ln - 4) / 8
            if q_field_cnt == 0:
                struct_str = '<%dQH' % (q_field_cnt, )
            else:
                struct_str = '<QH'
            counts = struct.unpack(struct_str, extra[offset:offset+ln])
```
with:
```
            if ln >= 28:
                counts = struct.unpack('<QQQ', extra[4:28])
            elif ln == 20:
                counts = struct.unpack('<QQ', extra[4:20])
            elif ln == 12:
                counts = struct.unpack('<Q', extra[4:12])
            elif ln == 4:
                counts = ()
            else:
                raise BadZipFile("Corrupt extra field %04x (size=%d)" % (EXTRA_ZIP64, ln))
```

Now it seems to work.

My OS: Windows 10 Pro x64 v2004 (Build 19041.264)
Python: 3.8.0
pyzipper: 0.3.1 (installed via pip)

Kind regards
Malte Hillmann

**Comment by danifus** [comment]
Thanks for the report.

I'll write a test and fix that up this week. I think I've switched the blocks of that if statement. It should be:

```
            q_field_cnt = (ln - 4) / 8
            if q_field_cnt == 0:
                struct_str = '<QH'
            else:
                struct_str = '<%dQH' % (q_field_cnt, )
            counts = struct.unpack(struct_str, extra[offset:offset+ln])
```

**Comment by danifus** [comment]
This is fixed in the 0.3.2 version.

Thanks for reporting the bug!
